docs/eeps/draft-spi.md

```diff
@@ -0,0 +1,251 @@
+
+The Embulk core had to expect that any class in the core could be called by some Embulk plugin because changing anything in the core could cause an incompatible change. This discouraged the maintainers from making changes to the Embulk core.
+
+Since v0.10, Embulk has clearly specified its [SPI (Service Provider Interface)](https://en.wikipedia.org/wiki/Service_provider_interface) for plugins, so that plugins would not have unnecessary dependencies on the details of the Embulk core implementation, and the maintainers could limit compatibility concerns. The SPI acts as a compatibility contract with plugins. We only care about compatibility with the SPI classes. The core classes outside the SPI may have incompatible changes at any time without notice.
+
+This was another issue related to [EEP-3: The "Compact Core" Principle](./eep-0003.md) inside the Embulk core codebase, in contrast to [EEP-7](./eep-0007.md), which was for dependency libraries.
+
+This EEP explains the details about the Embulk SPI separation.
+
+Separation by Maven Artifact
+=============================
+
+All the Embulk core classes had been included equally in the `embulk-core` Maven artifact. Every Embulk plugin implicitly depended on `embulk-core` like [the `provided` scope](https://maven.apache.org/guides/introduction/introduction-to-dependency-mechanism.html#Dependency_Scope). (See the example `build.gradle` below.)
+
+```gradle
+// An example from embulk-input-s3:v0.3.5
+// https://github.com/embulk/embulk-input-s3/blob/v0.3.5/build.gradle
+configurations {
+    provided
+    // ...
+}
+
+dependencies {
+    compile  "org.embulk:embulk-core:0.9.12"
+    provided "org.embulk:embulk-core:0.9.12"
+    // ...
+}
+
+task classpath(type: Copy, dependsOn: ["jar"]) {
+    // ...
+    from (configurations.runtime - configurations.provided + files(jar.archivePath))
+    // ...
+}
+```
+
+It was definitely hard for plugin developers to tell which class was officially provided for plugins, or not expected to be accessed by plugins, because they were equally accessible. It was also hard for the Embulk core maintainers to find an Embulk plugin accessing a core class that was not expected to be accessed by a plugin.
+
+We decided to have the `embulk-spi` Maven artifact split from `embulk-core` so that an Embulk plugin could explicitly depend only on the official Embulk SPI (`embulk-spi`). The core maintainers do not consider non-SPI classes need to be compatible.
+
+```gradle
+dependencies {
+    compileOnly "org.embulk:embulk-spi:0.11"
```

Comment:
This part used `compileOnly` not `compile`. And [the `provided` scope](https://maven.apache.org/guides/introduction/introduction-to-dependency-mechanism.html#Dependency_Scope) doesn't contain it. (Because It is gradle syntax).
It may help if a note like `compileOnly = compile` is added. (At least for me)

Reply:
Gradle's scopes (configurations) do not correspond to Maven's scopes one by one. Gradle's `compileOnly` is only Gradle's. It won't be present in the generated Maven `pom.xml`. `compileOnly ≠ compile`.

Here, `compileOnly "org.embulk:embulk-spi:0.11"` only means that Gradle includes `embulk-spi:0.11` in compilation, but does not include `embulk-spi:0.11` as runtime dependency.

Reply:
Thank you for your comment!